### PR TITLE
Cluster system tests for agent groups

### PR DESCRIPTION
**Pull request by yanazaeva**
|Related issue|
|---|
|https://github.com/wazuh/wazuh-qa/issues/2479|

## Description

This closes https://github.com/wazuh/wazuh-qa/issues/2479. The aim of this issue is to update the already existing system tests as well as create a new one to check the proper synchronization.

### CODE REVIEW

File: tests/reliability/test_cluster/test_cluster_logs/test_cluster_logs_order/data/Agent-groups_sync.yml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  parent: log4
+  tag: 'Finished in .* \(.* chunks updated\).*'
+
+# Race condition: 'All chunks sent' arrives (sometimes) after 'Finished in'.
+- log_id: log6
+  parent: log2
+  tag: 'Finished in .* \(.* chunks updated\).*'
+- log_id: log7
+  parent: log6
+  tag: 'All chunks sent.*'
```
Comment (AdriiiPRodri) [inline tests/reliability/test_cluster/test_cluster_logs/test_cluster_logs_order/data/Agent-groups_sync.yml:36]:
```suggestion

```

**Review by AdriiiPRodri — CHANGES_REQUESTED** [review]
We need to add tests to check the task master agent-groups get, master agent-groups send and worker agent-groups recv.

File: tests/reliability/test_cluster/test_cluster_logs/test_cluster_master_logs_order/test_cluster_master_logs_order.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    """
+    if not artifacts_path:
+        pytest.fail('Parameter "--artifacts_path=<path>" is required.')
+
+    cluster_log_files = os.path.join(artifacts_path, 'master', 'logs', 'cluster.log')
+
+    if len(cluster_log_files) == 0:
+        pytest.fail(f"No files found inside {artifacts_path}.")
+
+    all_workers = {'Master': LogsOrder().logs_order}
```
Comment (AdriiiPRodri) [inline tests/reliability/test_cluster/test_cluster_logs/test_cluster_master_logs_order/test_cluster_master_logs_order.py:66]:
```suggestion
    master = {'Master': LogsOrder().logs_order}
```

File: tests/reliability/test_cluster/test_cluster_logs/README.md [hunk 3]
```diff
@@ -39,14 +39,15 @@ collected 5 items
...
-========================================================================================= 4 passed in 3.50 seconds ==========================================================================================
+========================================================================================= 5 passed in 3.52 seconds ==========================================================================================
 ```
 
 ### Tests information
 
 | Number of tests | Time spent |
 |:--:|:--:|
-| 4 | 6s |
+| 5 | 4s |
```
Comment (AdriiiPRodri) [inline tests/reliability/test_cluster/test_cluster_logs/README.md:53]: Check time again

File: tests/system/test_cluster/test_agent_groups/data/add_messages.yml [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    timeout: 60
+  - regex: '.*Agent-groups recv.*There is no data requiring synchronization in the local database.*'
+    path: "/var/ossec/logs/cluster.log"
+    timeout: 60
+  - regex: '.*Agent-groups recv.*The checksum of both databases match.*'
+    path: "/var/ossec/logs/cluster.log"
+    timeout: 60
+  - regex: '.*Agent-groups recv.*Finished in.*Updated.*chunks.*'
+    path: "/var/ossec/logs/cluster.log"
+    timeout: 60
```
Comment (AdriiiPRodri) [inline tests/system/test_cluster/test_agent_groups/data/add_messages.yml:75]:
```suggestion
```

**Review by AdriiiPRodri — CHANGES_REQUESTED** [review]

File: tests/reliability/test_cluster/test_cluster_logs/test_cluster_master_logs_order/README.md [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+The test needs to receive one parameter (artifacts) in order to be run. If this parameter is not specified, the test will fail. The required parameter is:
+- `--artifacts_path`: Path where cluster logs can be found inside each worker folder. It should follow the structure below:
+    ```.
+    ├── worker_x
+    │   └── logs
+    │       └── cluster.log
+    ├── worker_y
+    │   └── logs
+    │       └── cluster.log
+    └── ...
```
Comment (AdriiiPRodri) [inline tests/reliability/test_cluster/test_cluster_logs/test_cluster_master_logs_order/README.md:27]: This is the master README

File: tests/reliability/test_cluster/test_cluster_logs/test_cluster_master_logs_order/README.md [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+## Overview 
+
+Check that logs in the cluster nodes are printed in the expected order.
+
+This test checks the order of the logs for the three cluster tasks (`agent-info sync`, `integrity check` and `integrity sync`) in the worker nodes. 
+
+## Objective
+
+To verify that:
+- The cluster does not skip any task.
```
Comment (AdriiiPRodri) [inline tests/reliability/test_cluster/test_cluster_logs/test_cluster_master_logs_order/README.md:12]:
```suggestion
- The cluster does not skip any task. The master node does not skip any cluster's task.
```

File: tests/reliability/test_cluster/test_cluster_logs/test_cluster_master_logs_order/README.md [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+        """
+        if not artifacts_path:
+            pytest.fail('Parameter "--artifacts_path=<path>" is required.')
+    
+        cluster_log_files = os.path.join(artifacts_path, 'master', 'logs', 'cluster.log')
+    
+        if len(cluster_log_files) == 0:
+            pytest.fail(f"No files found inside {artifacts_path}.")
+    
+        all_workers = {'Master': LogsOrder().logs_order}
```
Comment (AdriiiPRodri) [inline tests/reliability/test_cluster/test_cluster_logs/test_cluster_master_logs_order/README.md:67]:
```suggestion
        all_managers = {'Master': LogsOrder().logs_order}
```

File: tests/reliability/test_cluster/test_cluster_logs/test_cluster_master_logs_order/README.md [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+                                                                      tree_info['tree'].children(tree_info['node'])],
+                                                    'found_log': result.group(3)})
+>                           pytest.fail(f"[{incorrect_order[0]['name']}]"
+                                        f"\n - Log type: {incorrect_order[0]['log_type']}"
+                                        f"\n - Expected logs: {incorrect_order[0]['expected_logs']}"
+                                        f"\n - Found log: {incorrect_order[0]['found_log']}")
+E                           Failed: [CLUSTER-Workload_benchmarks_metrics_B77_manager_1]
+E                            - Log type: Agent-groups send
+E                            - Expected logs: ['Finished in.*chunks updated.*']
+E                            - Found log: Fittttnished in 0.004s (1 chunks updated).
```
Comment (AdriiiPRodri) [inline tests/reliability/test_cluster/test_cluster_logs/test_cluster_master_logs_order/README.md:104]: This log is correct?

**Review by AdriiiPRodri — CHANGES_REQUESTED** [review]

**Review by AdriiiPRodri — APPROVED** [review]

**Comment by yanazaeva** [comment]
The testing was done using the following branches:

|Wazuh branch| Wazuh-QA branch|
|:--:|:--:|
|[dev-10771-agent-groups-files-to-wazuh-db](https://github.com/wazuh/wazuh/tree/dev-10771-agent-groups-files-to-wazuh-db)|[feature/2479-cluster-system-tests-for-agent-groups](https://github.com/wazuh/wazuh-qa/tree/feature/2479-cluster-system-tests-for-agent-groups)|

**Review by AdriiiPRodri — APPROVED** [review]

File: tests/reliability/test_cluster/test_cluster_logs/test_cluster_worker_logs_order/data/Agent-groups_recv.yml [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+# Synchronization needed.
+- log_id: log6
+  parent: log2
+  tag: 'Checksum comparison failed. Attempt.*'
+- log_id: log7
+  parent: log6
+  tag: 'Finished in.*Updated.*chunks.*'
+
+# Synchronization needed.
```
Comment (Selutario) [inline tests/reliability/test_cluster/test_cluster_logs/test_cluster_worker_logs_order/data/Agent-groups_recv.yml:36]: Update this comment

File: tests/reliability/test_cluster/test_cluster_logs/test_cluster_worker_logs_order/data/Agent-groups_recv.yml [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+  parent: log0
+  tag: '.*chunks updated in wazuh-db in.*'
+- log_id: log2
+  parent: log1
+  tag: 'Obtained.*chunks of data in.*'
+
+# No need to synchronize.
+- log_id: log3
+  parent: log2
+  tag: 'There is no data requiring synchronization in the local database.*'
```
Comment (Selutario) [inline tests/reliability/test_cluster/test_cluster_logs/test_cluster_worker_logs_order/data/Agent-groups_recv.yml:20]: This is a debug2 log

**Review by Selutario — CHANGES_REQUESTED** [review]

File: tests/system/test_cluster/test_integrity_sync/test_integrity_sync.py [hunk 11]
```diff
@@ -171,12 +171,14 @@ def test_extra_files(clean_files):
...
 
-def test_extra_valid_files(clean_files):
+def not_test_extra_valid_files(clean_files):
     """Check that extra_valid files created in the workers are copied in the master.
 
     Register two agents, wait and check that the client.keys file have been updated in all worker nodes.
     Then, each worker creates an 'agent-groups' file for one of the created agents. Files inside
     /var/ossec/queue/agent-groups are 'extra_valid' according to cluster.json, so they should be sent to the master.
+
+    This test is no longer in use, but it could be useful in the future.
```
Comment (juliamagan) [inline tests/system/test_cluster/test_integrity_sync/test_integrity_sync.py:181]: We should delete this test if it is no longer in use. The test already exists in `4.3` if we need it in the future.

**Review by juliamagan — CHANGES_REQUESTED** [review]
GJ, but some changes are required

**Review by juliamagan — DISMISSED** [review]
LGTM

**Review by juliamagan — DISMISSED** [review]

**Comment by juliamagan** [comment]
## 16/03/2022
### Testing
|Wazuh branch| Wazuh-QA branch|
|:--:|:--:|
|[dev-10771-agent-groups-files-to-wazuh-db](https://github.com/wazuh/wazuh/tree/dev-10771-agent-groups-files-to-wazuh-db)|[feature/2479-cluster-system-tests-for-agent-groups](https://github.com/wazuh/wazuh-qa/tree/feature/2479-cluster-system-tests-for-agent-groups)|

#### tests/system/test_cluster/test_agent_groups/test_agent_groups.py
||Results|
|:--:|:--:|
|R1|[🟢 ](https://github.com/wazuh/wazuh-qa/files/8261470/R1-test-agent-groups.zip)
|R2|[🟢 ](https://github.com/wazuh/wazuh-qa/files/8261471/R2-test-agent-groups.zip)
|R3|[🟢 ](https://github.com/wazuh/wazuh-qa/files/8261472/R3-test-agent-groups.zip)

#### tests/system/test_cluster/test_agent_files_deletion/test_agent_files_deletion.py
||Results|
|:--:|:--:|
|R1|
|R2|
|R3|

#### tests/system/test_cluster/test_integrity_sync/test_integrity_sync.py
||Results|
|:--:|:--:|
|R1|
|R2|
|R3|

#### tests/reliability/test_cluster/test_cluster_logs/test_cluster_master_logs_order/test_cluster_master_logs_order.py
||Results|
|:--:|:--:|
|R1|
|R2|
|R3|

#### tests/reliability/test_cluster/test_cluster_logs/test_cluster_worker_logs_order/test_cluster_worker_logs_order.py
||Results|
|:--:|:--:|
|R1|
|R2|
|R3|

**Comment by yanazaeva** [comment]
## 25/03/2022
### Testing
|Wazuh branch| Wazuh-QA branch|
|:--:|:--:|
|[dev-10771-agent-groups-files-to-wazuh-db](https://github.com/wazuh/wazuh/tree/dev-10771-agent-groups-files-to-wazuh-db)|[feature/2479-cluster-system-tests-for-agent-groups](https://github.com/wazuh/wazuh-qa/tree/feature/2479-cluster-system-tests-for-agent-groups)|

#### tests/system/test_cluster/test_agent_groups/test_agent_groups.py
||Results|
|:--:|:--:|
|R1|[🟢 ](https://github.com/wazuh/wazuh-qa/files/8352707/1.zip)
|R2|[🟢 ](https://github.com/wazuh/wazuh-qa/files/8352708/2.zip)
|R3|[🟢 ](https://github.com/wazuh/wazuh-qa/files/8352709/3.zip)

#### tests/system/test_cluster/test_agent_files_deletion/test_agent_files_deletion.py
||Results|
|:--:|:--:|
|R1|[🟢 ](https://github.com/wazuh/wazuh-qa/files/8352789/1.zip)
|R2|[🟢 ](https://github.com/wazuh/wazuh-qa/files/8352790/2.zip)
|R3|[🟢 ](https://github.com/wazuh/wazuh-qa/files/8352791/3.zip)

#### tests/system/test_cluster/test_integrity_sync/test_integrity_sync.py
||Results|
|:--:|:--:|
|R1|
|R2|
|R3|

#### tests/reliability/test_cluster/test_cluster_logs/test_cluster_master_logs_order/test_cluster_master_logs_order.py
||Results|
|:--:|:--:|
|R1|
|R2|
|R3|

#### tests/reliability/test_cluster/test_cluster_logs/test_cluster_worker_logs_order/test_cluster_worker_logs_order.py
||Results|
|:--:|:--:|
|R1|
|R2|
|R3|

**Comment by yanazaeva** [comment]
## Update

We repeated the tests, with the pull mentioned above merged and with the following [artifacts](https://github.com/wazuh/wazuh-qa/files/8379792/artifacts.zip), which do not contain any restart. 

#### tests/reliability/test_cluster/test_cluster_logs/test_cluster_master_logs_order/test_cluster_master_logs_order.py
||Results|
|:--:|:--:|
|R1|[:green_circle: ](https://github.com/wazuh/wazuh-qa/files/8379775/1.zip)
|R2|[:green_circle: ](https://github.com/wazuh/wazuh-qa/files/8379776/2.zip)
|R3|[:green_circle: ](https://github.com/wazuh/wazuh-qa/files/8379777/3.zip)

#### tests/reliability/test_cluster/test_cluster_logs/test_cluster_worker_logs_order/test_cluster_worker_logs_order.py
||Results|
|:--:|:--:|
|R1|[🟢 ](https://github.com/wazuh/wazuh-qa/files/8379760/1.zip)
|R2|[🟢 ](https://github.com/wazuh/wazuh-qa/files/8379761/2.zip)
|R3|[🟢 ](https://github.com/wazuh/wazuh-qa/files/8379762/3.zip)

**Comment by Rebits** [comment]
## 06/4/2022

### Testing
|Wazuh branch| Wazuh-QA branch|
|:--:|:--:|
|[dev-10771-agent-groups-files-to-wazuh-db](https://github.com/wazuh/wazuh/tree/dev-10771-agent-groups-files-to-wazuh-db)|[feature/2479-cluster-system-tests-for-agent-groups](https://github.com/wazuh/wazuh-qa/tree/feature/2479-cluster-system-tests-for-agent-groups)|

### `test_cluster_master_logs_order`

|  OS  | Local        | Notes
|---    |---     |---    |
| R1   | [:green_circle: ](https://github.com/wazuh/wazuh-qa/files/8426644/R1-2645-test_cluster_master_logs_order.zip)  |  |
| R2   | [:green_circle:](https://github.com/wazuh/wazuh-qa/files/8426646/R2-2645-test_cluster_master_logs_order.zip)   |  |
| R3   | [:green_circle:](https://github.com/wazuh/wazuh-qa/files/8426648/R3-2645-test_cluster_master_logs_order.zip)   |   |


### test_cluster_worker_logs_order

|  OS  | Local   | Jenkins     | Notes
|---    |---    |---    |---    |
| R1   | [:green_circle: ](https://github.com/wazuh/wazuh-qa/files/8426645/R1-2645-test_cluster_worker_logs_order.zip)  |  |
| R2   | [:green_circle:](https://github.com/wazuh/wazuh-qa/files/8426647/R2-2645-test_cluster_worker_logs_order.zip)   |  |
| R3   | [:green_circle:](https://github.com/wazuh/wazuh-qa/files/8426649/R3-2645-test_cluster_worker_logs_order.zip)   |  |
* * * 

- :green_circle:: All pass
- :yellow_circle:: Some warnings
- :red_circle:: Some errors/fails
- :large_blue_circle:: In progress

**Review by Rebits — APPROVED** [review]

**Pull request merged**